lib/tpm2_alg_util.c

```diff
@@ -120,3 +123,169 @@ TPM_ALG_ID tpm2_alg_util_from_optarg(char *optarg) {
+        }
+
+        *digest_spec_str = '\0';
+        digest_spec_str++;
+
+        bool result = tpm2_util_string_to_uint32(pcr_index, &dspec->pcr_index);
+        if (!result) {
+            LOG_ERR("Got invalid PCR Index: \"%s\", in digest spec: \"%s\"",
+                    pcr_index, spec);
+            return false;
+        }
+
+        /* now that the pcr_index is removed, parse the remaining <hash_name>=<hash_value>,.. */
+        char *digest_hash_tok;
+        char *save_ptr = NULL;
+
+        /* keep track of digests we have seen */
+
+        while ((digest_hash_tok = strtok_r(digest_spec_str, ",", &save_ptr))) {
+            digest_spec_str = NULL;
+
+            if (count >= ARRAY_LEN(dspec->digests.digests)) {
+                LOG_ERR("Specified too many digests per spec, max is: %zu",
+                        ARRAY_LEN(dspec->digests.digests));
+                return false;
+            }
+
+            TPMT_HA *d = &dspec->digests.digests[count];
+
+            char *stralg = digest_hash_tok;
+            char *split = strchr(digest_hash_tok, '=');
+            *split = '\0';
+            split++;
+
+            char *data = split;
+
+            /*
+             * Convert and validate the hash algorithm. It should be a hash algorithm
+             */
+            TPM_ALG_ID alg = tpm2_alg_util_from_optarg(stralg);
+            if (alg == TPM_ALG_ERROR) {
+                LOG_ERR("Could not convert algorithm, got: \"%s\"", stralg);
+                return false;
+            }
+
+            bool is_hash_alg = tpm2_alg_util_is_hash_alg(alg);
+            if (!is_hash_alg) {
+                LOG_ERR("Algorithm is not a hash algorithm, got: \"%s\"",
+                        stralg);
+                return false;
+            }
+
+            d->hashAlg = alg;
+
+            /* fill up the TPMT_HA structure with algorithm and digest */
+            BYTE *digest_data = (BYTE *) &d->digest;
+
+            UINT16 expected_hash_size = hash_alg_size(alg);
+            /* strip any preceding hex on the data as tpm2_util_hex_to_byte_structure doesn't support it */
+            bool is_hex = !strncmp("0x", data, 2);
+            if (is_hex) {
+                data += 2;
+            }
+
+            UINT16 size =  expected_hash_size;
+            int rc = tpm2_util_hex_to_byte_structure(data, &size,
+                    digest_data);
+            if (rc) {
+                LOG_ERR("Error \"%s\" converting hex string as data, got:"
+                    " \"%s\"", hex_to_byte_err(rc), data);
+                return false;
+            }
+
+            if (expected_hash_size != size) {
+                LOG_ERR(
+                        "Algorithm \"%s\" expects a size of %u bytes, got: %u",
+                        stralg, expected_hash_size, size);
+                return false;
+            }
+
+            count++;
+        }
+
+        /* assign count at the end, so count is 0 on error */
+        dspec->digests.count = count;
```

Comment:
If it doesn't go through this loop, count = 0 should cause a failure. Ie missing digest hash value spec... or nothing found with ,